### PR TITLE
Fix v2 training with reaction data using CLI

### PR DESCRIPTION
**Pull request by xiaoruiDong**
This PR fixes the pipeline when training and predicting with reaction data using CLI:
- correct the argument `mode_` when creating `CGRFeaturizer`
- add a `__call__` function to `CGRFeaturizer` which is basically calling `featurize` 
- Move reaction smiles processing inside of `_ReactionDataPointMixin.from_smi`. A separate processing is included in `v2.cli.utils_.make_datapoints` previously.
- use `.from_smi` when creating `ReactionDatapoint` from SMILES
- correct how the molecule featurizer is fed into the `ReactionDatapoint` or `MoleculeDatapoint`. Previously, it is fed as a string where `*Datapoint`s ask for a list of featurizer objects. Now create the featurizer list in `make_datapoints`.
- add `__post_init__` to `ReactionDataset` similarly to `MoleculeDataset` to force assigning `.__Y` and `.__X_f`, so that `Y` and `X_f` properties can be queried. The implementation is consistent with `MoleculeDataset`.
- Use keyword argument assignment style when creating `MPNN` object in `v2.cli.train.main`. Previously, the argument assignment is misaligned.
- Change the data type of the morgan fingerprints generated from rdkit to `int32`. The default data type is `uint32`, however, PyTorch doesn't support converting numpy arrays of uint32 to `Tensor`. Use `int32` instead which should be able to cover all of the potential use cases.

The PR also includes a few relevant fixes or updates to several CLI arguments:
- `--no-header-row`. Fix the case where`target_cols` is fixed to `[1]` in predict.
- `--task-type`. Update the `add_argument` with `choices`.
- `--loss-function`. Update the `add_argument` with `choices` and `help`.
- `separate-test-*` or `separate-val-*`. Update the `add_argument` with `dest` as variable names without "separate" is actually used in later steps.

## Checklist
- [ ] linted with flake8?
- [ ] (if appropriate) unit tests added?

### CODE REVIEW

File: chemprop/v2/featurizers/reaction.py [hunk 1]
```diff
@@ -82,6 +82,18 @@ def mode(self) -> RxnMode:
...
+        atom_features_extra: np.ndarray | None = None,
+        bond_features_extra: np.ndarray | None = None,
+    ) -> MolGraph:
+        return self.featurize(
+            rxn=rxn,
+            atom_features_extra=atom_features_extra,
+            bond_features_extra=bond_features_extra,
+        )
+
     def featurize(
```
Comment (davidegraff) [inline chemprop/v2/featurizers/reaction.py:97]: I think the `featurize` method is a vestige of a different commit. Just rename `featurize` to `__call__`

File: chemprop/v2/data/datasets.py [hunk 2]
```diff
@@ -292,6 +292,12 @@ class ReactionDataset(Dataset, _MolGraphDatasetMixin):
...
 
+    def __post_init__(self):
+        if self.data is None:
+            raise ValueError("Data cannot be None!")
+
+        self.reset()
+
     def __getitem__(self, idx: int) -> Datum:
         d = self.data[idx]
         mg = self.featurizer(((d.rct, d.pdt)), None, None)
```
Comment (davidegraff) [inline chemprop/v2/data/datasets.py:303]:
```suggestion
        mg = self.featurizer((d.rct, d.pdt), None, None)
```

File: chemprop/v2/featurizers/featurizers.py [hunk 3]
```diff
@@ -24,12 +24,12 @@ def __len__(self) -> int:
 
 class BinaryFeaturizerMixin:
     def __call__(self, mol: Chem.Mol) -> np.ndarray:
-        return self.F.GetFingerprintAsNumPy(mol)
+        return self.F.GetFingerprintAsNumPy(mol).astype(np.int32)
```
Comment (davidegraff) [inline chemprop/v2/featurizers/featurizers.py:27]:
```suggestion
        return self.F.GetFingerprintAsNumPy(mol).astype(int)
```
use full precision for now as they'll get cast to half-precision later during batch collation

File: chemprop/v2/featurizers/featurizers.py [hunk 4]
```diff
@@ -24,12 +24,12 @@ def __len__(self) -> int:
...
 
 
 class CountFeaturizerMixin:
     def __call__(self, mol: Chem.Mol) -> np.ndarray:
-        return self.F.GetCountFingerprintAsNumPy(mol)
+        return self.F.GetCountFingerprintAsNumPy(mol).astype(np.int32)
```
Comment (davidegraff) [inline chemprop/v2/featurizers/featurizers.py:32]: see above

**Review by davidegraff — COMMENTED** [review]
just needs to merge in `v2/dev` and some small changes

**Comment by xiaoruiDong** [comment]
> just needs to merge in `v2/dev` and some small changes

@davidegraff Thanks for reviewing this PR! I have addressed the comments accordingly. The branch is already based on the latest v2/dev. It should be good to go.

File: chemprop/v2/cli/train.py [hunk 5]
```diff
@@ -252,10 +252,9 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
         "-t",
         "--task-type",
         default="regression",
-        action=RegistryAction(
```
Comment (KnathanM) [inline chemprop/v2/cli/train.py:255]: @davidegraff can you remind me why we use `action=RegistryAction(<registry>)` instead of `choices=<registry>.keys()`? Reading the source code, to me they look equivalent and Xiaorui's change here makes it more simple.

Comment (davidegraff) [reply chemprop/v2/cli/train.py:255]: the `RegistryAction` sets the `choices` argument automatically, but more importantly it essentially functions as a dictionary lookup. That is when a client calls `--task-type regression`, then `args.task_type` is automatically set to `ReadoutRegistry["regression"]` (which is a `RegressionFFN`).
perhaps (in another PR) we will want to change the `--task-type` `metavar` to something like `metavar="ffn"`

**Review by davidegraff — DISMISSED** [review]
👍  lgtm!

File: chemprop/v2/cli/train.py (hunk 5 above)
Comment (davidegraff) [reply chemprop/v2/cli/train.py:255]: yeah wait @xiaoruiDong why is this code changed? The `RegistryAction` should take care of both of these things.

File: chemprop/v2/cli/train.py [hunk 6]
```diff
@@ -615,16 +628,17 @@ def main(args):
...
+        agg=agg,
+        readout=readout_ffn,
+        batch_norm=True,
+        metrics=None,
+        w_t=args.task_weights,
+        warmup_epochs=args.warmup_epochs,
+        # num_lrs=args.num_lrs,  # this argument is currently commented out in MPNN.__init__()
+        init_lr=args.init_lr,
+        max_lr=args.max_lr,
+        final_lr=args.final_lr,
```
Comment (davidegraff) [inline chemprop/v2/cli/train.py:641]: delete keywords. We can't have every PR introducing stylistic changes of previously written code based on the author's preferences. Most of the library uses positional arguments, so it should reasonably maintain the convention where possible. I.e., here the keywords are exactly the same names as the variables.

File: chemprop/v2/cli/train.py [hunk 7]
```diff
@@ -615,16 +628,17 @@ def main(args):
         test_loader = None
 
     model = MPNN(
-        mp_block,
-        agg,
-        readout_ffn,
-        None,
-        args.task_weights,
-        args.warmup_epochs,
-        args.num_lrs,
```
Comment (davidegraff) [inline chemprop/v2/cli/train.py:624]: remove `args.num_lrs` if it's not in the `MPNN.__init__` signature.

File: chemprop/v2/cli/utils_.py [hunk 8]
```diff
@@ -115,26 +117,24 @@ def make_datapoints(
...
+                rxn_or_smis=smis[i],
+                keep_h=keep_h,
+                add_h=add_h,
+                y=targetss[i],
+                weight=weights[i],
+                gt_mask=gt_targetss[i],
+                lt_mask=lt_targetss[i],
+                x_f=featuress[i],
+                mfs=mfs,
+                x_phase=None,
```
Comment (davidegraff) [inline chemprop/v2/cli/utils_.py:135]: same here

File: chemprop/v2/cli/utils_.py [hunk 9]
```diff
@@ -154,7 +154,7 @@ def make_datapoints(
                 gt_mask=gt_targetss[i],
                 lt_mask=lt_targetss[i],
                 x_f=featuress[i],
-                mfs=features_generators,
+                mfs=mfs,
                 x_phase=None,
                 keep_h=keep_h,
                 add_h=add_h,
```
Comment (davidegraff) [inline chemprop/v2/cli/utils_.py:160]: change to positionals while you're at it.

**Review by davidegraff — CHANGES_REQUESTED** [review]

File: chemprop/v2/cli/train.py (hunk 6 above)
Comment (xiaoruiDong) [reply chemprop/v2/cli/train.py:641]: OK. The reason why I changed this is because I found since we were using positional arguments, we actually didn't realized that we misaligned the argument at all, starting from `batch_norm`. I can switch back to positional if you think that is proper.

File: chemprop/v2/cli/train.py (hunk 7 above)
Comment (xiaoruiDong) [reply chemprop/v2/cli/train.py:624]: `num_lrs` is currently commented out in `MPNN.__init__`. Do you know we will implement it in the future? If not, I can clean up the `MPNN.__init__` as well.

File: chemprop/v2/cli/train.py (hunk 5 above)
Comment (xiaoruiDong) [reply chemprop/v2/cli/train.py:255]: 1. Sorry that I mistakenly changed this to `ReadoutRegistry.keys()`... I found, in a recent merged PR, Hao-Wei did a similar change and thought this was the preferred way to do.
2. If I read @davidegraff correctly, do we expect to see `args.task_type == RegressionFFN` when someone do `--task-type regression`? Currently the actual behavior is that `args.task_type` is a str `regression` with `action=RegistryAction(ReadoutRegistry)`

**Comment by xiaoruiDong** [comment]
@davidegraff thank you so much for the detailed check again. I reverted the changes of `RegistryAction` and keyword arguments as suggested. I also removed `num_lrs` from `MPNN`s and updated unit tests for CGRfeaturizer.

File: chemprop/v2/cli/train.py (hunk 6 above)
Comment (KnathanM) [reply chemprop/v2/cli/train.py]: FWIW, I think we should have used a different convention, i.e. used keyword arguments when passing many (more than four?) arguments. Even when the keywords and variable names are the same, using positional arguments leads to (sometimes silent) errors when the arguments are misaligned.

File: chemprop/v2/cli/train.py (hunk 7 above)
Comment (KnathanM) [reply chemprop/v2/cli/train.py:624]: We have decided to not implement it in the future. Thank you Xiaorui.

File: chemprop/v2/cli/train.py (hunk 6 above)
Comment (davidegraff) [reply chemprop/v2/cli/train.py]: On the flip side, changing a variable name in a function definition can break a lot of client code if you're relying on keyword arguments.

File: chemprop/v2/cli/train.py (hunk 5 above)
Comment (davidegraff) [reply chemprop/v2/cli/train.py:255]: The second point was mistaken. I had it that way in an old version and I decided to change it back in a previous commit, but maybe we should change it back to avoid some boilerplate client code? In any case, you should be using the `RegistryAction`.

**Review by davidegraff — CHANGES_REQUESTED** [review]
I'm willing to compromise: arguments without defaults should use positionals and those with default values can use keywords

**Review by davidegraff — CHANGES_REQUESTED** [review]
also the signature of `ReactionDatapoint.from_smi` should changed to:
```python
@classmethod
    def from_smi(
        cls,
        rxn_or_smis: str | tuple[str, str],
        *args,
        keep_h: bool = False,
        add_h: bool = False,
        **kwargs,
    )
```
to align with `MoleculeDatapoint` (this way creating a datapoint from a SMILES vs a Mol is the same function signature more-or-less)

**Comment by KnathanM** [comment]
> I'm willing to compromise: arguments without defaults should use positionals and those with default values can use keywords

I didn't end up making this change as the positional arguments are at least known to be in the right positions now for the cases in this PR. 

With the last two commits, I believe this PR is ready to merge @davidegraff.

**Comment by xiaoruiDong** [comment]
@KnathanM thank you for helping me addressing David's comment!

**Review by davidegraff — APPROVED** [review]
sorry for forgetting about this 👍 (but please merge `v2/dev` in first)

**Comment by xiaoruiDong** [comment]
@davidegraff No worry. Thank you for reviewing! I merged in the latest `v2/dev`.
